File: 437-path-sum-iii/path-sum-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int count;
    private Map<Long, Long> map;
    private void solve(TreeNode root, long sum, long target) {
        if(root == null){
            return;
        }

        sum += root.val;
        if(map.containsKey(sum - target)){
            count += map.get(sum - target);
        }
        map.put(sum, map.getOrDefault(sum, 0L) + 1);

        solve(root.left, sum, target);
        solve(root.right, sum, target);

        map.put(sum, map.getOrDefault(sum, 0L) - 1);
    }

    public int pathSum(TreeNode root, int targetSum) {
        count = 0;
        map = new HashMap<>();
        map.put(0L, 1L);
        solve(root, 0, targetSum);
        return count;
    }
}

File: 437-path-sum-iii/path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    int count=0;
    int target;
    unordered_map<long long,long long> map;
    void dfs(TreeNode* root,long long prefixSum)
    {
        if(root==NULL)
        return;

        prefixSum+=root->val;

        if(map.count(prefixSum-target)>0)    
        count+=map[prefixSum-target];

        map[prefixSum]++;

        dfs(root->left,prefixSum);
        dfs(root->right,prefixSum);

        // backtracking
        map[prefixSum]--;
        if(map[prefixSum]==0)
        map.erase(prefixSum);
    }

    int pathSum(TreeNode* root, int targetSum) 
    {
        if(root==NULL)
        return 0;

        target=targetSum;
        map[0]=1;
        
        dfs(root,0);
        return count;
    }
};